### PR TITLE
add unit testing json read number

### PR DESCRIPTION
**Pull request by ghost**
# Description

The goal here was to implement additional testing so that the JSON_unit_test also checks that it can read different number types, other sections already test the INT reading so this just tests the double and float reading in order to make sure the number is correct and the type is correct

The test section is Json can be created and read with different number types, in the test case of json can be created and read.

## Type of change

- [✓] Adding additional tests

## How Has This Been Tested?

In order to test this pull request you have to make the project, the cd into bin and run the command "./skunit_tests [json] -s" in your MSYS2 MINGW64 terminal, this will give you a read out of the unit test, the section labelled "Json can be created and read with different number types" will contain the changes made for the reading of float and double.

additionally you can test it by just running "./skunit_tests [json]" but this wont give you the same readout about the actual tests being conducted. 

"json_read_number" should the number as a float and "json_read_number_as_double" should read it as double and the number should be the number same as the one set into the JSON

## Testing Checklist

- [✓] Tested with skunit_tests

## Checklist

- [✓ ] My code follows the style guidelines of this project
- [ ✓] I have performed a self-review of my own code
- [ ✓] I have commented my code in hard-to-understand areas
- [ ✓] I have made corresponding changes to the documentation
- [ ✓] My changes generate no new warnings
- [ ✓] I have requested a review from ... on the Pull Request

### CODE REVIEW

**Review by JPF2209 — APPROVED** [review]
# Important
- this code is good enough to approve but before the 2nd peer review, please change line 120 with Json in capitals since it doesn't line up with all other json's in the document and consistency is crucial for anything in SplashKit

## Type of Change
- This piece of code correctly identifies the changes made being adding additional tests

## Code Readability
- The code is very understandable and clear in what it needs to be comparing it to the other code.

## Maintainability
- As this is in the same style as the other code, it's quite maintainable being simple while doing the job.

## Code Simplicity
- This code is simple in it's execution following established design patterns and best practices 

## Edge Cases
- The code deals with edge cases simply since the inner functionality of the test code manages this

## Test Thoroughness
- For all the key tests in this type of scenario, this covers it all and the code has been tested in multiple environments and it works.

## Backward Compatibility
- The code doesn't break existing functionality in the way it works.

## Performance Considerations
- Performance is usually more of a concern with the function itself but this works out well for performance.

## Security Concerns
- This code has no impact negatively or positively for security.

## Dependencies
- There are no new added dependencies. 

## Documentation
- The documentation provided is through and simple to follow at the same time

**Comment by ghost** [comment]
following the previous peer review I have changed the Json capitilization to json

File: coresdk/src/test/unit_tests/unit_test_json.cpp [hunk 1]
```diff
@@ -117,7 +117,18 @@ TEST_CASE("json can be created and read", "[json]")
 
         REQUIRE("#00ff00ff" == color_to_string(deserialized_clr));
     }
+    SECTION("json can be created and read with different number types")
```
Comment (dijidiji) [inline coresdk/src/test/unit_tests/unit_test_json.cpp]: This could be its own `TEST_CASE` since you're creating a new json object rather than reusing the `person` json created in the parent case.
```suggestion
TEST_CASE("json can be created and read with different number types", "[json_read_number][json_read_number_as_double]")
```
Alternatively, you could reuse the previous json object.
In this case, the test will run pretty similarly (the code in the containing TEST_CASE or SECTION will run once for each nested SECTION), so it's mostly for organisation/readability. It's worth noting in case you want to take advantage of nested SECTIONs.

File: coresdk/src/test/unit_tests/unit_test_json.cpp [hunk 2]
```diff
@@ -117,7 +117,18 @@ TEST_CASE("json can be created and read", "[json]")
...
+    {
+        json number_types = create_json();
+
+        json_set_number(number_types, "float", static_cast<float>(21.2));
+        json_set_number(number_types, "double", static_cast<double>(30.1));
```
Comment (dijidiji) [inline coresdk/src/test/unit_tests/unit_test_json.cpp]: You can write floating point literals without a cast here to make it a bit easier to read:

```suggestion
        json_set_number(number_types, "float", 21.2f);
        json_set_number(number_types, "double", 30.1);
```
[There's a more in depth look at the C++ reference docs](https://en.cppreference.com/w/cpp/language/floating_literal.html)

File: coresdk/src/test/unit_tests/unit_test_json.cpp [hunk 3]
```diff
@@ -117,7 +117,18 @@ TEST_CASE("json can be created and read", "[json]")
...
 
+        REQUIRE(json_read_number(number_types, "float") == static_cast<float>(21.2));
+        REQUIRE(json_read_number_as_double(number_types, "double") == static_cast<double>(30.1));
```
Comment (dijidiji) [inline coresdk/src/test/unit_tests/unit_test_json.cpp]: For floating point comparisons it'd be good if we use matchers to avoid issues with different floating point representations of a number not being exactly equal. There are a few different kinds ([see Catch2 docs for more info](https://github.com/catchorg/Catch2/blob/v2.x/docs/matchers.md)), but I've been using WithinRelMatcher with the default epsilon, which is covenient.
You can use this by adding a using declaration for the relevant matcher and then using the `REQUIRE_THAT(value/object, matcher expression)` syntax. For example, for the WithinRelMatcher:
```cpp
using Catch::Matchers::WithinRel;
```
```suggestion
        REQUIRE_THAT(json_read_number(number_types, "float"), WithinRel(21.2f));
        REQUIRE_THAT(json_read_number_as_double(number_types, "double"), WithinRel(30.1));
```
Note the floating-point literal suffix f here, too.

**Review by dijidiji — CHANGES_REQUESTED** [review]
Just chiming in with some useful pointers for unit testing! I'll have to make a note of this in the unit testing guide.

**Comment by ghost** [comment]
thank you for the suggestions, sorry about the delay github didn't notify me of them for some reason. The changes have been applied and they all work successfully with no issues.

**Review by dijidiji — CHANGES_REQUESTED** [review]
Sorry, experienced a similar issue to you and didn't receive a GitHub notification. To use `WithinRel` you'll need to add the relevant using directive (preferably after the includes at the top of the file):
```cpp
using Catch::Matchers::WithinRel;
```

**Comment by rory-cd** [comment]
This PR has been continued [here](https://github.com/thoth-tech/splashkit-core/pull/142).

**Review by monicavtasmin — APPROVED** [review]
